frame/pools/src/lib.rs

```diff
@@ -0,0 +1,863 @@
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+pub struct RewardPool<T: Config> {
+	/// The balance of this reward pool after the last claimed payout.
+	balance: BalanceOf<T>,
+	/// The points of this reward pool after the last claimed payout
+	points: BalanceOf<T>,
+	/// The total earnings _ever_ of this reward pool after the last claimed payout. I.E. the sum
+	/// of all incoming balance through the pools life.
+	total_earnings: BalanceOf<T>,
+	/// The reward destination for the pool.
+	account_id: T::AccountId,
+}
+
+impl<T: Config> RewardPool<T> {
+	fn update_total_earnings_and_balance(mut self) -> Self {
+		let current_balance = T::Currency::free_balance(&self.account_id);
+		// The earnings since the last time it was updated
+		let new_earnings = current_balance.saturating_sub(self.balance);
+		// The lifetime earnings of the of the reward pool
+		self.total_earnings = new_earnings.saturating_add(self.total_earnings);
+		self.balance = current_balance;
+
+		self
+	}
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, DefaultNoBound)]
+#[cfg_attr(feature = "std", derive(frame_support::DebugNoBound, Clone, PartialEq))]
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+struct UnbondPool<T: Config> {
+	points: PointsOf<T>,
+	balance: BalanceOf<T>,
+}
+
+impl<T: Config> UnbondPool<T> {
+	fn points_to_issue(&self, new_funds: BalanceOf<T>) -> PointsOf<T> {
+		if self.balance.is_zero() || self.points.is_zero() {
+			debug_assert!(self.balance.is_zero() && self.points.is_zero());
+
+			new_funds.saturating_mul(POINTS_TO_BALANCE_INIT_RATIO.into())
+		} else {
+			let points_per_balance = {
+				let balance = T::BalanceToU128::convert(self.balance);
+				let points = T::BalanceToU128::convert(self.points);
+				// REMINDER: `saturating_from_rational` panics if denominator is zero
+				FixedU128::saturating_from_rational(points, balance)
+			};
+			let new_funds = T::BalanceToU128::convert(new_funds);
+
+			T::U128ToBalance::convert(points_per_balance.saturating_mul_int(new_funds))
+		}
+	}
+
+	fn balance_to_unbond(&self, delegator_points: PointsOf<T>) -> BalanceOf<T> {
+		if self.balance.is_zero() || delegator_points.is_zero() {
+			// There is nothing to unbond
+			return Zero::zero()
+		}
+
+		let balance_per_share = {
+			let balance = T::BalanceToU128::convert(self.balance);
+			let points = T::BalanceToU128::convert(self.points);
+			// REMINDER: `saturating_from_rational` panics if denominator is zero
+			FixedU128::saturating_from_rational(balance, points)
+		};
+		let delegator_points = T::BalanceToU128::convert(delegator_points);
+
+		T::U128ToBalance::convert(balance_per_share.saturating_mul_int(delegator_points))
+	}
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, DefaultNoBound)]
+#[cfg_attr(feature = "std", derive(frame_support::DebugNoBound, Clone, PartialEq))]
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+struct SubPoolsContainer<T: Config> {
```

Comment:
opinionated: not a fan of calling something that's (IMO clearly) a container (because of the plural 's') a `container`.

Reply:
The naming is hard because I don't want to get it confused with the `SubPools` storage map. I would take suggestion for better names, but I just don't want the storage map and the item in the storage map to have the same name

Reply:
Fair enough and end of the day your opinion is more important here, but for that, I generally recommend:

```
struct Foo { .. }
#[pallet::storage]
type FooStorage = StorageVallue<_, Foo>; 
```